ARROW-13548: [C++] Implement temporal difference kernels

lidavidm opened the pull request:
Implements kernels to get the number of {years, months, days and millis, days, hours, minutes, seconds, millis, micros, nanos} between two timestamps. Everything returns either Int64 or DayTimeInterval.

lidavidm commented:
I think Python doesn't have bindings for the interval types - that should be fixed in a separate issue I would think.

ianmcook commented:
Would it be straightforward to add a week variant of this that takes `DayOfWeekOptions`?

lidavidm commented:
> Would it be straightforward to add a week variant of this that takes `DayOfWeekOptions`?

It should be doable. I'll update the PR.

nealrichardson commented:
@rok can you take a look here?

lidavidm commented:
I'll rebate this in a little while

rok reviewed (commented):
Thanks @lidavidm !
Couple of comments :)

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -830,6 +897,167 @@ struct ISOCalendar {
+      : week_start_(options->week_start), localizer_(std::move(localizer)) {}
+
+  template <typename T, typename Arg0, typename Arg1>
+  T Call(KernelContext*, Arg0 arg0, Arg1 arg1, Status*) const {
```
Perhaps you could find the first week starting weekday after the starting date and then calculate the difference // 7?

lidavidm commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -192,6 +195,60 @@ struct TemporalComponentExtract
+  const auto& timezone = GetInputTimezone(batch.values[0]);
+  for (int i = 1; i < batch.num_values(); i++) {
+    const auto& other_timezone = GetInputTimezone(batch.values[i]);
+    if (other_timezone != timezone) {
+      return Status::TypeError("Got differing time zone '", other_timezone,
+                               "' for argument ", i + 1, "; expected '", timezone, "'");
+    }
```
What would it mean to find the difference between two timestamps in different timezones? I suppose we could compute the difference in UTC, but then you'd get different results (as compared to working with local timestamps) if you had a DST transition in the covered time.

rok replied:
Difference between two moments is just an interval without a timezone. And out timezone aware timestamps are unambiguous moments or am I getting something terribly wrong here? :)

lidavidm replied:
Oh. Doh. Yes, you are right. Thanks for pointing this out :)

lidavidm replied:
Er, wait. So concretely, if I have these timestamps:

```
2019-12-31 18:00:00-0500
2019-12-31 23:00:00-0500
```

In America/New_York, the year difference is 0, but in UTC, the year difference is 1 (the latter timestamp is 04:00 on 2020-01-01). This kernel will localize the timestamps, so the expected difference is 0. But this kind of calculation doesn't make sense if the time zones are different, unless you suddenly switch to interpreting zoned timestamps as their underlying UTC value. I'm worried this is potentially confusing/a footgun.

rok replied:
Riiight it's because you care about boundaries not the full periods :). In that case you'd want to block this indeed and have user change one of the two timestamps into the timezone they care about and run the operation 'there'.

rok replied:
Sorry for the misdirect.

rok replied:
Sorry for the misdirect.

lidavidm replied:
Ok, sounds good, sorry for the confusion. I realized I forgot to update the docs, I'll try to add some explanatory text there as well since this is a little confusing.

lidavidm commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -830,6 +897,167 @@ struct ISOCalendar {
+        std::chrono::duration_cast<std::chrono::milliseconds>(to - floor<days>(to))
+            .count());
+    const int32_t num_millis = to_time - from_time;
+    return DayTimeIntervalType::DayMilliseconds{num_days, num_millis};
```
You mean, why didn't I add kernels for MonthsBetween and MonthDayNano between? I can add those.

rok replied:
Feels like it would be good for completeness.

lidavidm commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1104,6 +1375,90 @@ const FunctionDoc assume_timezone_doc{
     "AssumeTimezoneOptions"};
 
+const FunctionDoc years_between_doc{
+    "Compute the number of year boundaries between two timestamps",
```
How about this?

```
const FunctionDoc years_between_doc{
    "Compute the number of years between two timestamps",
    ("Returns the number of year boundaries crossed from the first timestamp to the\n"
     "second. That is, the difference is calculated as if the timestamps were\n"
     "truncated to the year.\n"
     "Null values emit null."),
    {"start", "end"}};
```

rok replied:
Better. How about using 'complete years passed/elapsed/..' instead of 'year boundaries crossed'?

lidavidm replied:
I suppose the reason why I'm using 'boundaries' is because the difference between 1998-12-01 and 1999-01-1 is '1 year' here even though a complete year hasn't passed.

rok replied:
Oh, I thought the entire year needs to pass. Right, boundaries are better yeah.

lidavidm commented:
Changes:
- Simplified the week kernel (thanks for the suggestion)
- Added docs
- Added kernels to get differences as month interval and as month-day-nano interval
Note the interval types are not bound in Python, see ARROW-14018.

rok commented:
I'll take another look tomorrow.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1103,6 +1399,115 @@ const FunctionDoc assume_timezone_doc{
 
+const FunctionDoc years_between_doc{
+    "Compute the number of years between two timestamps",
+    ("Returns the number of year boundaries crossed from the first timestamp to the\n"
```
Should `\n`s appear mid sentence?

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1103,6 +1399,115 @@ const FunctionDoc assume_timezone_doc{
 
+const FunctionDoc years_between_doc{
+    "Compute the number of years between two timestamps",
+    ("Returns the number of year boundaries crossed from the first timestamp to the\n"
```
Should `\n`s appear mid sentence? Same below.

lidavidm replied:
I think it's fine, e.g. see https://github.com/apache/arrow/blob/5260fd5970d34c62e0ae7680f606b29eab56d96b/cpp/src/arrow/compute/kernels/vector_nested.cc#L141 or https://github.com/apache/arrow/blob/5260fd5970d34c62e0ae7680f606b29eab56d96b/cpp/src/arrow/compute/kernels/vector_sort.cc#L1643-L1650

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1103,6 +1399,115 @@ const FunctionDoc assume_timezone_doc{
+     "timestamps were truncated to the months, then the  then the difference between "
+     "time times\n"
+     "of the two timestamps is computed as if both times were truncated to the\n"
+     "millisecond.\n"
```
```suggestion
     "nanosecond.\n"
```

rok replied:
Same here.

rok commented on docs/source/cpp/compute.rst:
Sometimes you use `_interval_` and sometimes you don't. I realize it's due to the type returned but it still bothers my eyes :). Probably best keep it as is.

lidavidm replied:
Yeah, I wasn't sure how to disambiguate these. Note that month_interval_between and months_between are basically redundant, except MonthIntervalType is int32_t, not int64_t.

pitrou replied:
Well, is it useful to have a `int64_t` return?

lidavidm replied:
Not particularly, except for consistency; that said for Python, we don't have the interval types bound yet, so there'll be a period where this kernel is unavailable if we remove this version. I think overall it's not worth keeping it though so I'll remove it.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1103,6 +1399,115 @@ const FunctionDoc assume_timezone_doc{
+
+const FunctionDoc month_day_nano_interval_between_doc{
+    "Compute the number of months, days and nanoseconds between two timestamps",
+    ("Returns the number of months, days, and milliseconds from the first timestamp\n"
```
Did this get lost?

lidavidm replied:
Fixed this (and the one below), I think I forgot to pull before pushing.

rok replied:
Happens to me all the time :)

rok commented:
LGTM @lidavidm!
Perhaps it would be good to ask for another review - I might be missing something.

pitrou commented on cpp/src/arrow/util/formatting_util_test.cc:
```diff
@@ -427,4 +427,42 @@ TEST(Formatting, Timestamp) {
+  {
+    StringFormatter<MonthIntervalType> formatter(month_interval());
+
+    AssertFormatting(formatter, 0, "0m");
```
Is it a recommended format for months? Should we use "M" instead?

lidavidm replied:
Actually, I suppose that doesn't make any sense. M seems preferable to m though so I'll do that.

pitrou commented on docs/source/cpp/compute.rst:
If the "notes" column is empty, then can omit it.

pitrou commented on docs/source/cpp/compute.rst:
"has a defined timezone" is less weird than "is zoned".

pitrou commented on docs/source/cpp/compute.rst:
I would say "because the local year is the same even though the UTC years would be different".

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal_test.cc:
```diff
@@ -433,6 +512,173 @@ TEST_F(ScalarTemporalTest, DayOfWeek) {
+  DayOfWeekOptions options(/*one_based_numbering=*/false, /*week_start=Monday*/ 1);
+  EXPECT_THAT(CallFunction("weeks_between",
+                           {
+                               *MakeArrayFromScalar(
```
Is `MakeArrayFromScalar` necessary or is the kernel able to broadcast a scalar input?

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -191,6 +198,60 @@ struct TemporalComponentExtract
+  return Status::OK();
+}
+
+Status ValidateDayOfWeekOptions(const DayOfWeekOptions& options) {
```
At some point, we may instead define a `Validate()` method on FunctionOptions.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -191,6 +198,60 @@ struct TemporalComponentExtract
+
+template <template <typename...> class Op, typename Duration, typename InType,
+          typename OutType>
+struct TemporalComponentExtractBinary {
```
Nit, but "Extract" refers to component extraction, which this isn't doing?

pitrou commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1103,6 +1399,114 @@ const FunctionDoc assume_timezone_doc{
 
+const FunctionDoc years_between_doc{
+    "Compute the number of years between two timestamps",
+    ("Returns the number of year boundaries crossed from the first timestamp to the\n"
```
How about: ```Return the number of year boundaries crossed from `start` to `end`.```

lidavidm commented:
I've rebased this.

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -776,6 +840,188 @@ struct ISOCalendar {
+  QuartersBetween(const FunctionOptions* options, Localizer&& localizer)
+      : localizer_(std::move(localizer)) {}
+
+  static int64_t GetQuarters(const year_month_day& ymd) {
+    return static_cast<int64_t>(static_cast<int32_t>(ymd.year())) * 4 +
+           (static_cast<uint32_t>(ymd.month()) - 1) / 3;
+  }
```
You could use this function for the quarter extractor and delete this logic there.

rok replied:
I mean this: https://github.com/apache/arrow/blob/7124872e99673d1087e7c547babf0fb3dd2634d7/cpp/src/arrow/compute/kernels/scalar_temporal.cc#L408

rok reviewed (commented):
Looks great @lidavidm !

Do we want to support this for time/date/32/64 at some point?

lidavidm commented:
> Looks great @lidavidm !
> 
> Do we want to support this for time/date/32/64 at some point?

Thanks for the review. Yeah, let me add support for the other temporal types - I suppose all that's merged by now.

lidavidm commented:
I've added the support for date/time types (years, months, etc. don't accept time types).

rok commented on docs/source/cpp/compute.rst:
```suggestion
+---------------------------------+------------+-------------------+-----------------------+----------------------------+
| quarters_between                | Binary     | Timestamp, Date   | Int64                 |                            |
+---------------------------------+------------+-------------------+-----------------------+----------------------------+
```

rok commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1209,6 +1660,75 @@ void RegisterScalarTemporal(FunctionRegistry* registry) {
+          &default_day_of_week_options, DayOfWeekState::Init);
+  DCHECK_OK(registry->AddFunction(std::move(weeks_between)));
+
+  auto day_time_between =
```
Nit: `day_time_interval_between` would be more consistent. Same for `month_day_nano_interval_between`. I don't feel strongly about it, just noticing :).

rok reviewed (approved):
Looks good!

lidavidm commented:
Thanks, fixed those issues.

rok commented:
I think this is ready to go.
